# Analyse des données de transport

## Objectifs

Ce notebool a pour objectif d'explorer et d'analyser un historique de tickets de transports:
* Analyser les prix et durée des trajets
* comparer les modes de transport selon la distance
* identifier des tendances et incohérences dans les données
* proposer des analyses et visualisations pertinentes

## Outils et bibliothèques utilisés
* Python
* pandas
* numpy
* matplotlib
* scikit-learn
* pyarrow
* pyjanitor
* ipykernel
* jupyter
* seaborn
* polars
* plotly
* folium
* missingno
* great_expectations

## Auteur
Haja Rabemananjara

## Date
17 janvier 2026

## 1. Import des librairies et configuration de l'apparence globale des graphiques

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
pd.set_option('display.precision', 2)

## 2. Chargement des données

In [282]:
ticket = pd.read_csv('data/ticket_data.csv')
cities = pd.read_csv('data/cities.csv')
stations = pd.read_csv('data/stations.csv')
providers = pd.read_csv('data/providers.csv')

###### Aperçu des données

In [283]:
print("Dimensions des datasets:")
{ticket.shape, cities.shape, stations.shape, providers.shape}

Dimensions des datasets:


{(227, 10), (8040, 6), (11035, 4), (74168, 12)}

In [284]:
from IPython.display import display, Markdown

display(Markdown("### Info sur les tickets"))
ticket.head()
ticket.info()

display(Markdown("### Info sur les villes"))
cities.head()
cities.info()

display(Markdown("### Info sur les stations"))
stations.head()
stations.info()

display(Markdown("### Info sur les fournisseurs"))
providers.head()
providers.info()

### Info sur les tickets

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74168 entries, 0 to 74167
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               74168 non-null  int64  
 1   company          74168 non-null  int64  
 2   o_station        32727 non-null  float64
 3   d_station        32727 non-null  float64
 4   departure_ts     74168 non-null  object 
 5   arrival_ts       74168 non-null  object 
 6   price_in_cents   74168 non-null  int64  
 7   search_ts        74168 non-null  object 
 8   middle_stations  32727 non-null  object 
 9   other_companies  32727 non-null  object 
 10  o_city           74168 non-null  int64  
 11  d_city           74168 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 6.8+ MB


### Info sur les villes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8040 entries, 0 to 8039
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           8040 non-null   int64  
 1   local_name   8040 non-null   object 
 2   unique_name  8039 non-null   object 
 3   latitude     8040 non-null   float64
 4   longitude    8040 non-null   float64
 5   population   369 non-null    float64
dtypes: float64(3), int64(1), object(2)
memory usage: 377.0+ KB


### Info sur les stations

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11035 entries, 0 to 11034
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           11035 non-null  int64  
 1   unique_name  11035 non-null  object 
 2   latitude     11035 non-null  float64
 3   longitude    11035 non-null  float64
dtypes: float64(2), int64(1), object(1)
memory usage: 345.0+ KB


### Info sur les fournisseurs

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 227 entries, 0 to 226
Data columns (total 10 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   id                    227 non-null    int64 
 1   company_id            227 non-null    int64 
 2   provider_id           213 non-null    object
 3   name                  227 non-null    object
 4   fullname              227 non-null    object
 5   has_wifi              224 non-null    object
 6   has_plug              224 non-null    object
 7   has_adjustable_seats  224 non-null    object
 8   has_bicycle           224 non-null    object
 9   transport_type        227 non-null    object
dtypes: int64(2), object(8)
memory usage: 17.9+ KB


In [285]:
print("Description statistique des tickets")
ticket.describe()

Description statistique des tickets


id   company  o_station  d_station  price_in_cents    o_city  \
count  7.42e+04  74168.00   32727.00   32727.00        74168.00  74168.00   
mean   6.83e+06   7109.57    2907.13    2347.86         4382.71    849.19   
std    2.14e+04   3005.38    3347.63    3090.80         3739.33   1485.79   
min    6.80e+06      9.00       3.00       3.00          300.00      5.00   
25%    6.81e+06   8376.00     400.00     396.00         1900.00    485.00   
50%    6.83e+06   8385.00     701.00     575.00         3350.00    628.00   
75%    6.85e+06   8385.00    6246.00    4538.00         5250.00    628.00   
max    6.87e+06   8387.00   11017.00   11017.00        38550.00  12190.00   

         d_city  
count  74168.00  
mean     883.78  
std     1654.70  
min        1.00  
25%      453.00  
50%      562.00  
75%      628.00  
max    12190.00

## 3. Compréhension métier des données

### Description des colonnes clés
* *price_in_cents* : prix du ticket en cents (€)
* *departure_ts / arrival_ts* : départ et arrivée : format data YYYY-mm-jj HH:mm:ss
* *o_city / d_city* : ville de départ / arrivée
* *o_station / d_station* : station de départ / arrivée
* *company* : provider (ex : TGV, TER, Blablacar)

## 4. Nettoyage de données

#### Valeurs manquantes

In [286]:
print('Vérification des valeurs manquantes dans les tickets')
ticket.isna().sum()

Vérification des valeurs manquantes dans les tickets


id                     0
company                0
o_station          41441
d_station          41441
departure_ts           0
arrival_ts             0
price_in_cents         0
search_ts              0
middle_stations    41441
other_companies    41441
o_city                 0
d_city                 0
dtype: int64

On constate que 44 % (41441 / 74168) n'ont pas d'information sur les stations ('o_station' , 'd_station').
Cela correspond probablement aux trajets en covoiturage qui ne passent pas par des stations fixes.

In [287]:
missing_data = pd.DataFrame({
    'Dataset': ['Ticket', 'Cities', 'Stations', 'Providers'],
    'Total Rows': [len(ticket), len(cities), len(stations), len(providers)],
    'Missing Values': [ticket.isnull().sum().sum(), cities.isnull().sum().sum(), stations.isnull().sum().sum(), providers.isnull().sum().sum()]
})
print("Résumé des valeurs manquantes")
missing_data

Résumé des valeurs manquantes


Dataset  Total Rows  Missing Values
0     Ticket       74168          165764
1     Cities        8040            7672
2   Stations       11035               0
3  Providers         227              26

Détail des valeurs manquantes dans le dataset des tickets:"

In [288]:
missing_ticket = ticket.isnull().sum()
missing_ticket_pct = ((missing_ticket / len(ticket)) * 100).round(2)
missing_ticket_df = pd.DataFrame({
    'Valeurs manquantes': missing_ticket,
    'Pourcentage (%)': missing_ticket_pct
})
missing_ticket_df = missing_ticket_df[missing_ticket_df['Valeurs manquantes'] > 0]
missing_ticket_df.sort_values('Valeurs manquantes', ascending=False)

Valeurs manquantes  Pourcentage (%)
o_station                     41441            55.87
d_station                     41441            55.87
middle_stations               41441            55.87
other_companies               41441            55.87

#### Doublons

La suppression des doublons permettra d'éviter que les résultats soient biaisés.

In [289]:
ticket = ticket.drop_duplicates()

if ticket.drop_duplicates().shape[0] == ticket.shape[0]:
    print("Aucun doublon trouvé.")
else:
    print('Après suppression des doublons :', ticket.shape)

Aucun doublon trouvé.


#### Conversion

#### Conversion des prix en euros

In [290]:
ticket['price_eur'] = ticket['price_in_cents'] / 100
ticket[['price_in_cents', 'price_eur']].head()

price_in_cents  price_eur
0            4550       45.5
1            1450       14.5
2            7400       74.0
3           13500      135.0
4            7710       77.1

#### Conversion des timestamps

In [291]:
ticket['departure_ts'] = pd.to_datetime(ticket['departure_ts'], format='mixed')
ticket['arrival_ts'] = pd.to_datetime(ticket['arrival_ts'], format='mixed')
ticket['search_ts'] = pd.to_datetime(ticket['search_ts'], format='mixed')

Calcul de la durée des trajets en minutes

In [292]:
ticket['duration_min'] = (ticket['arrival_ts'] - ticket['departure_ts']).dt.total_seconds() / 60
ticket[['departure_ts', 'arrival_ts', 'duration_min']].head()

departure_ts                arrival_ts  duration_min
0 2017-10-13 14:00:00+00:00 2017-10-13 20:10:00+00:00         370.0
1 2017-10-13 13:05:00+00:00 2017-10-14 06:55:00+00:00        1070.0
2 2017-10-13 13:27:00+00:00 2017-10-14 21:24:00+00:00        1917.0
3 2017-10-13 13:27:00+00:00 2017-10-14 11:02:00+00:00        1295.0
4 2017-10-13 21:46:00+00:00 2017-10-14 19:32:00+00:00        1306.0

Extraction de features temporelles

In [293]:
ticket['departure_hour'] = ticket['departure_ts'].dt.hour
ticket['departure_day'] = ticket['departure_ts'].dt.day_name()
ticket['departure_month'] = ticket['departure_ts'].dt.month_name()
ticket['is_weekend'] = ticket['departure_ts'].dt.dayofweek.isin([5, 6])

Délai de réservation en jours

In [294]:
ticket['booking_ts'] = (ticket['departure_ts'] - ticket['search_ts']).dt.days

ticket[['search_ts', 'departure_ts', 'booking_ts']].head()

search_ts              departure_ts  booking_ts
0 2017-10-01 00:13:31.327000+00:00 2017-10-13 14:00:00+00:00          12
1 2017-10-01 00:13:35.773000+00:00 2017-10-13 13:05:00+00:00          12
2 2017-10-01 00:13:40.212000+00:00 2017-10-13 13:27:00+00:00          12
3 2017-10-01 00:13:40.213000+00:00 2017-10-13 13:27:00+00:00          12
4 2017-10-01 00:13:40.213000+00:00 2017-10-13 21:46:00+00:00          12

#### Valeurs abérantes

Vérification de l'existance de données abérantes.
Comparaison de la forme des tableau avant et après filtrage des données abérantes.
Suppression des valeurs abérantes si nécessaire.

Nettoyage des enregistrements avec prix négatif ou nul

In [295]:
print('Avant nettoyage :', ticket.shape)
ticket_filter = ticket[(ticket['price_in_cents'] > 0)]
print('Après nettoyage :', ticket_filter.shape)

if ticket.shape[0] == ticket_filter.shape[0]:
    print("Aucun enregistrement supprimé.")
else:
    ticket = ticket[(ticket['price_in_cents'] > 0)]

Avant nettoyage : (74168, 19)
Après nettoyage : (74168, 19)
Aucun enregistrement supprimé.


Nettyage des enregistrements avec durée négative ou nulle

In [296]:
print('Avant nettoyage :', ticket.shape)
ticket_filter = ticket[(ticket['duration_min'] > 0)]
print('Après nettoyage :', ticket_filter.shape)

if ticket.shape[0] == ticket_filter.shape[0]:
    print("Aucun enregistrement supprimé.")
else:
    ticket = ticket[(ticket['duration_min'] > 0)]

Avant nettoyage : (74168, 19)
Après nettoyage : (74168, 19)
Aucun enregistrement supprimé.


## 5. Jointures des datasets

Jointure progressive (vile -> stations -> fournisseurs)

### Jointure des données des tickets avec les villes d'origine.

In [297]:
cities_origin = cities[['id', 'local_name', 'latitude', 'longitude']].copy()
cities_origin = cities_origin.rename(columns={
    'id': 'city_id',
    'local_name': 'o_city_name',
    'latitude': 'o_lat',
    'longitude': 'o_lon'
})

"""Merge des données des tickets avec les villes d'origine."""
ticket = ticket.merge(cities_origin, left_on='o_city', right_on='city_id', how='left').drop(columns=['city_id'])


### Jointure des données des tickets avec les villes de destinations

In [298]:
cities_destination = cities[['id', 'local_name', 'latitude', 'longitude']].copy()
cities_destination = cities_destination.rename(columns={
    'id': 'city_id',
    'local_name': 'd_city_name',
    'latitude': 'd_city_latitude',
    'longitude': 'd_city_longitude'
})

"""Merge des données des tickets avec les villes de destinations."""
ticket = ticket.merge(cities_destination, left_on='d_city', right_on='city_id', how='left').drop(columns=['city_id'])

### Jointure des données avec les stations d'origine

In [299]:
stations_origin = stations[['id', 'unique_name', 'latitude', 'longitude']].copy()
stations_origin = stations_origin.rename(columns={
    'id': 'station_id',
    'unique_name': 'o_station_name',
    'latitude': 'o_station_latitude',
    'longitude': 'o_station_longitude'
})

"""Merge des données avec les stations d'origine."""
ticket = ticket.merge(stations_origin, left_on='o_station', right_on='station_id', how='left').drop(columns=['station_id'])

### Jointure des données avec les stations de destination

In [300]:
stations_destination = stations[['id', 'unique_name', 'latitude', 'longitude']].copy()
stations_destination = stations_destination.rename(columns={
    'id': 'station_id',
    'unique_name': 'd_station_name',
    'latitude': 'd_station_latitude',
    'longitude': 'd_station_longitude'
})

"""Merge des données avec les stations de destination."""
ticket = ticket.merge(stations_destination, left_on='d_station', right_on='station_id', how='left').drop(columns=['station_id'])

### Jointure des données avec les fournisseurs

In [301]:

ticket = ticket.merge(providers[['id', 'name', 'transport_type', 'has_wifi', 'has_plug', 'has_bicycle']], left_on='company', right_on='id', how='left', suffixes=('', '_provider')).drop(columns=['id_provider']).rename(columns={'name': 'provider_name'})

Colonnes finales du dataset:

In [302]:
print(ticket.columns.tolist())

print('Aperçu des données enrichies:')
ticket[['o_city_name', 'd_city_name', 'provider_name', 'transport_type', 'price_eur', 'duration_min']].head(10)

['id', 'company', 'o_station', 'd_station', 'departure_ts', 'arrival_ts', 'price_in_cents', 'search_ts', 'middle_stations', 'other_companies', 'o_city', 'd_city', 'price_eur', 'duration_min', 'departure_hour', 'departure_day', 'departure_month', 'is_weekend', 'booking_ts', 'o_city_name', 'o_lat', 'o_lon', 'd_city_name', 'd_city_latitude', 'd_city_longitude', 'o_station_name', 'o_station_latitude', 'o_station_longitude', 'd_station_name', 'd_station_latitude', 'd_station_longitude', 'provider_name', 'transport_type', 'has_wifi', 'has_plug', 'has_bicycle']
Aperçu des données enrichies:


o_city_name                     d_city_name  \
0  Orléans, Centre-Val de Loire, France  Montpellier, Occitanie, France   
1  Orléans, Centre-Val de Loire, France  Montpellier, Occitanie, France   
2  Orléans, Centre-Val de Loire, France  Montpellier, Occitanie, France   
3  Orléans, Centre-Val de Loire, France  Montpellier, Occitanie, France   
4  Orléans, Centre-Val de Loire, France  Montpellier, Occitanie, France   
5          Paris, Île-de-France, France  Lille, Hauts-de-France, France   
6          Paris, Île-de-France, France  Lille, Hauts-de-France, France   
7          Paris, Île-de-France, France  Lille, Hauts-de-France, France   
8          Paris, Île-de-France, France  Lille, Hauts-de-France, France   
9          Paris, Île-de-France, France  Lille, Hauts-de-France, France   

     provider_name transport_type  price_eur  duration_min  
0              bbc     carpooling       45.5         370.0  
1           ouibus            bus       14.5        1070.0  
2  corailintercite          train       74.0        1917.0  
3  corailintercite          train      135.0        1295.0  
4      coraillunea          train       77.1        1306.0  
5              bbc     carpooling       18.0         180.0  
6              bbc     carpooling       21.5         150.0  
7              bbc     carpooling       17.0         150.0  
8              bbc     carpooling       17.0         170.0  
9              bbc     carpooling       19.0         170.0

## 6. Calcul de la distance

### Calcul de la distance pour chaque trajet
Utilisation de la **formule de Haversine** pour calculer la distance entre deux coordonnées GPS

In [303]:
from math import radians, cos, sin, asin, sqrt

def haversine(lon1, lat1, lon2, lat2):
    """
    Calcule la distance en kilomètres entre deux point GPS
    en utilisant la formule de Haversine.
    """

    # gestion des valeurs manquantes
    if pd.isna(lon1) or pd.isna(lat1) or pd.isna(lon2) or pd.isna(lat2):
        return np.nan
    
    # conversion des degrés en radians 
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])

    # Formule de Haversine
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    km = 6371  # Rayon de la Terre en kilomètres
    return c * km

Calcul de la distance pour chaque trajet

In [304]:
ticket['distance_km'] = ticket.apply(lambda row: haversine(row['o_lon'], row['o_lat'], row['d_city_longitude'], row['d_city_latitude']), axis=1)
print("Statistiques descriptives de la distance en km:")
print(ticket['distance_km'].describe())

Statistiques descriptives de la distance en km:
count    74168.00
mean       363.04
std        194.93
min         18.91
25%        205.84
50%        338.32
75%        480.41
max       1870.17
Name: distance_km, dtype: float64


### Création des catégories de distance

In [305]:
ticket['distance_range'] = pd.cut(ticket['distance_km'], bins=[0, 200, 800, 2000, 10000], labels=['0-200 km', '201-800 km', '801-2000 km', '2000+ km)'])
print("Répartition des tickets par tranche de distance:")
print(ticket['distance_range'].value_counts().sort_index())

Répartition des tickets par tranche de distance:
distance_range
0-200 km       13724
201-800 km     58877
801-2000 km     1567
2000+ km)          0
Name: count, dtype: int64


### Calcul de la vitesse moyenne

In [306]:
ticket['vitesse_kmh'] = (ticket['distance_km'] / (ticket['duration_min'] / 60))

print("Vitesse moyenne par type de transport:")
print(ticket.groupby('transport_type')['vitesse_kmh'].mean().sort_values(ascending=False))

Vitesse moyenne par type de transport:
transport_type
train         75.57
carpooling    75.04
bus           35.77
Name: vitesse_kmh, dtype: float64


## 7. Analyse statistiques

### 7.1 Statistiques générales

#### Prix

In [307]:
print(f"Prix minimum: {ticket['price_eur'].min():.2f} €")
print(f"Prix moyen: {ticket['price_eur'].mean():.2f} €")
print(f"Prix maximum: {ticket['price_eur'].max():.2f} €")

Prix minimum: 3.00 €
Prix moyen: 43.83 €
Prix maximum: 385.50 €


#### Durée

In [308]:
print(f"Durée minimale: {ticket['duration_min'].min():.0f} minutes ({ticket['duration_min'].min()/60:.1f}h)")
print(f"Durée moyenne: {ticket['duration_min'].mean():.0f} minutes ({ticket['duration_min'].mean()/60:.1f}h)")
print(f"Durée maximale: {ticket['duration_min'].max():.0f} minutes ({ticket['duration_min'].max()/60:.1f}h)")


Durée minimale: 20 minutes (0.3h)
Durée moyenne: 425 minutes (7.1h)
Durée maximale: 29571 minutes (492.9h)


#### Distance

In [309]:
print(f"Distance minimale: {ticket['distance_km'].min():.0f} km")
print(f"Distance moyenne: {ticket['distance_km'].mean():.0f} km")
print(f"Distance maximale: {ticket['distance_km'].max():.0f} km")


Distance minimale: 19 km
Distance moyenne: 363 km
Distance maximale: 1870 km


### 7.2 Analyse par Type de Transport

In [311]:
transport_stats = ticket.groupby('transport_type').agg({
    'price_eur': ['min', 'mean', 'max'],
    'duration_min': ['min', 'mean', 'max'],
    'distance_km': ['min', 'mean', 'max'],
    'vitesse_kmh': ['mean']
}).round(2)

transport_stats.columns = ['_'.join(col).strip() for col in transport_stats.columns.values]
print(transport_stats)

                price_eur_min  price_eur_mean  price_eur_max  \
transport_type                                                 
bus                       8.5           36.52          229.0   
carpooling                3.0           27.42          161.5   
train                     4.9           85.07          385.5   

                duration_min_min  duration_min_mean  duration_min_max  \
transport_type                                                          
bus                         65.0             938.00           29571.0   
carpooling                  20.0             246.64            1750.0   
train                       39.0             440.06            2907.0   

                distance_km_min  distance_km_mean  distance_km_max  \
transport_type                                                       
bus                       27.38            474.42          1870.17   
carpooling                18.91            306.68          1754.30   
train                     19.28  

### 7.3 Analyse par Distance et Type de Transport

Prix et durée par distance et par type de transport

In [313]:
analysis_by_distance = ticket.groupby(['distance_range', 'transport_type']).agg({
    'price_eur': ['min', 'mean', 'max'],
    'duration_min': ['min', 'mean', 'max'],
    'vitesse_kmh': ['mean']
}).round(2)
print(analysis_by_distance)

                              price_eur                duration_min           \
                                    min    mean    max          min     mean   
distance_range transport_type                                                  
0-200 km       bus                 8.50   21.83  229.0         65.0   613.95   
               carpooling          3.00   11.77  128.5         20.0   117.12   
               train               4.90   34.88  251.0         39.0   265.16   
201-800 km     bus                10.00   34.95  224.8        149.0   904.49   
               carpooling          8.50   32.18  138.0         80.0   285.79   
               train              14.00   91.25  385.5         68.0   460.66   
801-2000 km    bus                22.90   69.50  174.0        735.0  1652.17   
               carpooling         44.96   86.27  161.5        470.0   786.01   
               train              19.40  154.64  375.5        213.0   729.52   
2000+ km)      bus                  NaN 

/tmp/ipykernel_7274/3709863832.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  analysis_by_distance = ticket.groupby(['distance_range', 'transport_type']).agg({
